harness/determined/_info.py

```diff
@@ -0,0 +1,318 @@
+        trial_seed: int,
+        hparams: Dict[str, Any],
+        config: Dict[str, Any],
+        latest_batch: int,
+        trial_run_id: int,
+        debug: bool,
+        unique_port_offset: int,
+        inter_node_network_interface: Optional[str],
+    ):
+        self.trial_id = trial_id
+        self.experiment_id = experiment_id
+        self.trial_seed = trial_seed
+        self.hparams = hparams
+
+        # _config is private because it's not a stable API; as the experiment config version
+        # increases, the shape of the experiment config dict might change as the master is upgraded.
+        # As a result, we should strongly discourage users from reading the experiment config dict
+        # in their training code.  They should focus on the data field and the hparams field, which
+        # have always been and will always be user-defined.
+        self._config = config
+
+        # rb: These fields are private because I am pretty confident that we need to find better
+        # ways to pass them around the system.  But for now, they're passed in as environment
+        # variables and for now we have to be able to handle that.
+        # TODO: decide if we really want to track latest_batch for users or not.
+        self._latest_batch = latest_batch
+        # TODO: get rid of trial_run_id in favor of allocation id.
+        self._trial_run_id = trial_run_id
+        # TODO: decide if the experiment config is the right place for users to set a debug flag.
+        self._debug = debug
+        # TODO: is this derivable from the slot ids?
+        self._unique_port_offset = unique_port_offset
+        # TODO: Get rid of this in favor of launch layer configs?
+        self._inter_node_network_interface = inter_node_network_interface
+
+    @classmethod
+    def _from_env(cls) -> "TrialInfo":
+        assert "DET_TRIAL_ID" in os.environ, "must be run inside a Trial container"
+        experiment_config = json.loads(os.environ["DET_EXPERIMENT_CONFIG"])
+        return cls(
+            trial_id=int(os.environ["DET_TRIAL_ID"]),
+            experiment_id=int(os.environ["DET_EXPERIMENT_ID"]),
+            trial_seed=int(os.environ["DET_TRIAL_SEED"]),
+            hparams=json.loads(os.environ["DET_HPARAMS"]),
+            config=experiment_config,
+            latest_batch=int(os.environ["DET_LATEST_BATCH"]),
+            trial_run_id=int(os.environ["DET_TRIAL_RUN_ID"]),
+            debug=experiment_config.get("debug", False),
+            unique_port_offset=int(os.environ["DET_UNIQUE_PORT_OFFSET"]),
+            inter_node_network_interface=os.environ.get("DET_INTER_NODE_NETWORK_INTERFACE"),
+        )
+
+    def _to_file(self, path: str = "/run/determined/info/trial.json") -> None:
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
+
+    @classmethod
+    def _from_file(cls, path: str = "/run/determined/info/trial.json") -> Optional["TrialInfo"]:
```

Comment:
what do you think about using pathlib/Path type for these? if nothing else, makes checking .exists() really straightforward

Reply:
> makes checking .exists() really straightforward

`os.path.exists()` is equally straight-forward.  Personally I think `pathlib.Path` is silly, but if you feel strongly I can use it here.  It's not user facing so it doesn't really matter.